### PR TITLE
`azurerm_key_vault` - support for Virtual Network Rules

### LINKED ISSUE #1491: Storage account within subnet creation failing - Cannot proceed with operation because subnets default-subnet of the virtual network ...are not provisioned. They are in Updating state..
Problem when creating 2 separate subnets with storage account endpoints following with creation of 2 separate storage account each connected to separate subnet.

SubnetsNotProvisioned:Cannot proceed with operation because subnets default-subnet of the virtual network ...are not provisioned. They are in Updating state.."Terraform does not automatically rollback in the face of errors.

### Terraform Version

v0.11.7

### Affected Resource(s)

azurerm_subnet
azurerm_storage_account

### Terraform Configuration Files

```
##################################################################################
# PROVIDERS
##################################################################################

provider "azurerm" {
  subscription_id = "${var.arm_subscription_id}"
  client_id       = "${var.arm_principal}"
  client_secret   = "${var.arm_password}"
  tenant_id       = "${var.tenant_id}"
  alias           = "default"
  version         = "~> 1.7"
}

##################################################################################
# RESOURCES
##################################################################################

resource "azurerm_resource_group" "network_rg" {
  name     = "projectNetwork${terraform.workspace}Ne-rg"
  location = "${local.workspace["location_project"]}"
  provider = "azurerm.default"
  tags     = "${local.common_tags}"
}

resource "azurerm_resource_group" "acr_rg" {
  name     = "projectAcr${terraform.workspace}Ne-rg"
  location = "${local.workspace["location_project"]}"
  provider = "azurerm.default"
  tags     = "${local.common_tags}"
}

resource "azurerm_resource_group" "bar_rg" {
  name     = "projectbar${terraform.workspace}Ne-rg"
  location = "${local.workspace["location_project"]}"
  provider = "azurerm.default"
  tags     = "${local.common_tags}"
}

resource "azurerm_resource_group" "aks_rg" {
  name     = "projectAks${terraform.workspace}Ne-rg"
  location = "${local.workspace["location_project"]}"
  provider = "azurerm.default"
  tags     = "${local.common_tags}"
}

resource "azurerm_resource_group" "monitor_rg" {
  name     = "projectMonitor${terraform.workspace}Ne-rg"
  location = "${local.workspace["location_project"]}"
  provider = "azurerm.default"
  tags     = "${local.common_tags}"
}

resource "azurerm_resource_group" "data_rg" {
  name     = "projectData${terraform.workspace}Ne-rg"
  location = "${local.workspace["location_project"]}"
  provider = "azurerm.default"
  tags     = "${local.common_tags}"
}

resource "azurerm_virtual_network" "project_vnet" {
  name                = "project${terraform.workspace}Ne-vnet"
  address_space       = ["${local.workspace["project_vnet_address_space"]}"]
  location            = "${azurerm_resource_group.network_rg.location}"
  resource_group_name = "${azurerm_resource_group.network_rg.name}"
  provider            = "azurerm.default"
  tags                = "${local.common_tags}"
}

resource "azurerm_subnet" "project_vnet_default_subnet" {
  name                 = "${local.workspace["project_vnet_default_subnet_name"]}"
  virtual_network_name = "${azurerm_virtual_network.project_vnet.name}"
  address_prefix       = "${local.workspace["project_vnet_default_subnet_address_prefix"]}"
  service_endpoints   = ["Microsoft.Storage"]
  resource_group_name = "${azurerm_resource_group.network_rg.name}"
  provider            = "azurerm.default"
}

resource "azurerm_subnet" "project_vnet_aks_subnet" {
  name                 = "${local.workspace["project_vnet_aks_subnet_name"]}"
  virtual_network_name = "${azurerm_virtual_network.project_vnet.name}"
  address_prefix       = "${local.workspace["project_vnet_aks_subnet_address_prefix"]}"
  service_endpoints   = ["Microsoft.Storage"]
  resource_group_name = "${azurerm_resource_group.network_rg.name}"
  provider            = "azurerm.default"
}

resource "azurerm_storage_account" "foo_sa" {
  name                      = "foodb${lower(terraform.workspace)}ne"
  account_tier              = "Standard"
  account_replication_type  = "${local.workspace["foo_sa_replication_type"]}"
  enable_https_traffic_only = true
  location                  = "${azurerm_resource_group.data_rg.location}"
  resource_group_name       = "${azurerm_resource_group.data_rg.name}"
  provider                  = "azurerm.default"
  tags                      = "${local.common_tags}"

  depends_on = [
    "azurerm_virtual_network.project_vnet",
    "azurerm_subnet.project_vnet_aks_subnet",
    "azurerm_subnet.project_vnet_default_subnet"
  ]

  network_rules {
    bypass                     = ["AzureServices"]
    ip_rules                   = ["${local.workspace["kingston_pip"]}"]
    virtual_network_subnet_ids = ["${azurerm_subnet.project_vnet_aks_subnet.id}"]
  }
}

resource "azurerm_storage_account" "bar_sa" {
  name                     = "barfiles${lower(terraform.workspace)}ne"
  account_tier             = "Standard"
  account_replication_type = "${local.workspace["bar_sa_replication_type"]}"
  location            = "${azurerm_resource_group.bar_rg.location}"
  resource_group_name = "${azurerm_resource_group.bar_rg.name}"
  provider            = "azurerm.default"
  tags                = "${local.common_tags}"

  depends_on = [
    "azurerm_virtual_network.project_vnet",
    "azurerm_subnet.project_vnet_default_subnet",
    "azurerm_subnet.project_vnet_aks_subnet",
  ]

  network_rules {
    bypass                     = ["AzureServices"]
    ip_rules                   = ["${local.workspace["some_pip"]}"]
    virtual_network_subnet_ids = ["${azurerm_subnet.project_vnet_default_subnet.id}"]
  }
}

```

### Debug Output

* azurerm_storage_account.bar_sa: Error waiting for Azure Storage Account "<account_name>" to be created: azure#WaitForCompletion: the number of retries has been exceeded: StatusCode=400 -- Original Error: Code="NetworkAclsValidationFailure" Message="Validation of network acls failure: SubnetsNotProvisioned:Cannot proceed with operation because subnets default-subnet of the virtual network /subscriptions/<subscription_id>/resourceGroups/<resource group name>/providers/Microsoft.Network/virtualNetworks/<network name> are not provisioned. They are in Updating state.."Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.

### Expected Behavior

Terraform should create:

2 subnets each with storage service endpoint
2 storage accounts each within separate subnet

### Actual Behavior

2 subnets each with storage service endpoint - are getting created
1 storage accounts is getting created and 2nd account is failing with error provided earlier

### Steps to Reproduce

run provided configuration

### Important Factoids

It works when I make one storage account dependant on the other

add to section:
resource "azurerm_storage_account" "foo_sa"

  depends_on = [
    "azurerm_virtual_network.project_vnet",
    "azurerm_subnet.project_vnet_aks_subnet",
    "azurerm_subnet.project_vnet_default_subnet",
    "azurerm_storage_account.bar_sa",
  ]

### References

* #0000

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This PR does four things:

* [x] dependencies: upgrades to `2018-02-14` of the Key Vault API's
* [x] Data Source `azurerm_key_vault` - exposing the `network_acls` block
* [x] `azurerm_key_vault` - adding support for setting `network_acls` (including Virtual Network Rules)
* [x] `azurerm_storage_account` - locking on the Virtual Network name when deleting the storage account

This should fix #1491 - but I've been unable to reproduce it

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
Key Vault Resource tests pass:

screenshot 2018-10-05 at 22 58 00

Storage Tests pass (ignoring some known test failures):

screenshot 2018-10-05 at 22 58 05

**Comment by tombuildsstuff** [comment]
Data Source tests pass:

screenshot 2018-10-06 at 00 05 41

File: azurerm/resource_arm_storage_account.go [hunk 1]
```diff
@@ -730,11 +731,46 @@ func resourceArmStorageAccountDelete(d *schema.ResourceData, meta interface{}) e
...
+	read, err := client.GetProperties(ctx, resourceGroup, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(read.Response) {
+			return nil
+		}
+
+		return fmt.Errorf("Error retrieving Storage Account %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	virtualNetworkNames := make([]string, 0)
```
Comment (katbyte) [inline azurerm/resource_arm_storage_account.go]: quick comment explaining why we are locking here might be handy

Comment (tombuildsstuff) [reply azurerm/resource_arm_storage_account.go]: done 👍🏻

**Pull request merged**